### PR TITLE
Journal entry: Ad Hoc Exploration

### PR DESCRIPTION
**Pull request by MikeWasson**
connects to #165 

For review - don't merge yet.

### CODE REVIEW

**Comment by MikeWasson** [comment]
btw, after I complete #241, this will become journal entry **3**, so I named the image folder accordingly.

File: docs/journal/02-exploring-recent-events.md [hunk 1]
```diff
@@ -1,7 +1,122 @@
...
+
+## Fabrikam requirements
+
+:memo: For more about our Fabrikam scenario, see [Introducing the Journey](00-introducing-the-journey.md). 
+
+Here are the specific requirements for our scenario:
+
+- Fabrikam needs to query temperature data by building or room number, over a specified time window. For example, "What was the average temperature in building 25 over the past hour?"
+
+**[What is the maximum time window that we are targeting? i.e., past day? past week?]**
```
Comment (bennage) [inline docs/journal/02-exploring-recent-events.md]: Given the current rates of event data, we can say that data from the 3 months will be available to query.

File: docs/journal/02-exploring-recent-events.md [hunk 2]
```diff
@@ -1,7 +1,122 @@
...
+## Solutions
+
+We explored the following solutions:
+
+- Azure Stream Analytics + Azure SQL
+- Elastic Search
+
+> **Time series databases** are another interesting possibility. As the name implies, time series  databases are optimized for time series data, which seems like a natural fit for IoT. We have not investigated this approach yet. 
+
+> We also looked at writing the raw events directly to SQL, PowerBI, or DocumentDB. However, none of these has enough throughput and/or capacity for our requirements. (See issues [88], [89], [90].)
```
Comment (bennage) [inline docs/journal/02-exploring-recent-events.md]: We can qualify this further in a later revision.

File: docs/journal/02-exploring-recent-events.md [hunk 3]
```diff
@@ -1,7 +1,122 @@
...
+**Advantages:**
+
+- Easy to configure. Stream Analytics uses a declarative model to specify the input and output streams, and the transformations to be performed by the processing.
+- Easy to manage. Stream Analytics is an Azure service, so you don't need to provision or maintain VMs. 
+- Stream Analytics can pull in *reference data* from blob storage. Reference data is static or infrequently changing data. For example, in our Fabrikam scenario, we use this look up the building number from the device ID.   
+- Stream Analytics is integrated with Event Hubs, which we picked for event ingestion. (See [Journal entry 1][journal1].)
+
+**Concerns: **
+
+- You have to configure the processing ahead of time. This limits the idea of "ad hoc" exploration somewhat, because the user is running queries against the aggregated results, not the raw event data.
```
Comment (bennage) [inline docs/journal/02-exploring-recent-events.md]: > This limits the idea of "ad hoc" exploration somewhat

That's a bit of an understatement. :wink:

**Comment by bennage** [comment]
I think that this is a great start. I'd like to talk more about the _aggregate in real-time_ vs the _aggregate in the query_, but we can do that in a later revision.

File: docs/journal/02-exploring-recent-events.md [hunk 4]
```diff
@@ -1,7 +1,122 @@
...
+- Queries typically involve a time range: “What was the average temperature during this 20-minute window?” 
+- Queries might need relational data: “What was the average temperature across all sensors in building 25?” 
+- Relational data might be stored separately from event data; for example, a device registry that maps device IDs to buildings.
+- Ad hoc exploration *tends* to target more recent data, rather than the entire set of historical data. The exact definition of “recent” depends on your particular scenario. It might mean the past 24 hours, the past week, or the past month.
+- The system must aggregate results into a meaningful view. IoT solutions can generate a *lot* of data, even across a relatively small time slice, so you need a way for a human to understand the results.
+- Ad hoc exploration is generally "user initiated," meaning the user submits a query (or perhaps logs into a dashboard), versus having the system detect a condition and notify the user.
+
+> We don't include critical alerts under ad hoc exploration. Ad hoc queries should execute quickly, but not necessarily in real time. (See the [real-time processing][real-time-milestone] milestone.)
```
Comment (francischeung) [inline docs/journal/02-exploring-recent-events.md]: We might want to differentiate from "Hot" data such as within 5 min of the event occurance. Ad hoc exploration may not be able to access data this young: indexing and latency time could be more than 5 min.

File: docs/journal/02-exploring-recent-events.md [hunk 5]
```diff
@@ -1,7 +1,122 @@
...
+
+> **Time series databases** are another interesting possibility. As the name implies, time series  databases are optimized for time series data, which seems like a natural fit for IoT. We have not investigated this approach yet. 
+
+> We also looked at writing the raw events directly to SQL, PowerBI, or DocumentDB. However, none of these has enough throughput and/or capacity for our requirements. (See issues [88], [89], [90].)
+
+### Stream Analytics + SQL
+
+[Azure Stream Analytics] is a real-time stream processing service. It uses an SQL-like language to process incoming data. Unlike traditional SQL, Stream Analytics runs queries over an event stream, making it easy to analyze events within a temporal window.     
+
+In this approach, we use Stream Analytics to aggregate the event data in real time and store the results in SQL. The aggregated result is small enough that SQL can handle the load.
```
Comment (francischeung) [inline docs/journal/02-exploring-recent-events.md]: "small" meaning number of results per second, not necessarily the amount of data.

File: docs/journal/02-exploring-recent-events.md [hunk 6]
```diff
@@ -1,7 +1,122 @@
...
+- **Ease of use**. The [Kibana] dashboard provides an easy way to query and visualize the data.
+- **Performance**. Elasticsearch can perform fast queries against large data sets. 
+- **Scalability**. Elasticsearch uses a distributed document store, so it scales horizontally. 
+- **Data fidelity**. This approach lets you query over the raw event data, rather than aggregating it first. 
+
+**Concerns:**
+
+- **Maintenance**. Elasticsearch is not an Azure service, so you must set up VMs, install and configure Elasticsearch, monitor the health of the cluster, and so forth. ([This blog post][elasticsearch-on-azure] has a nice write up of the steps.)
+- **Security**. Elasticsearch does not provide any security out of the box. You must run it inside a virtual network. 
+- **Not relational**. Elasticsearch is not relational and cannot use reference data (that is, data from another store). To add extra information such as the building number, the event processor must add it to the event data.
```
Comment (francischeung) [inline docs/journal/02-exploring-recent-events.md]: Another approach I've heard is to have a process that pulls data from Elasticsearch, adds the reference data, and then updates Elasticsearch.

File: docs/journal/02-exploring-recent-events.md (hunk 4 above)
Comment (bennage) [inline docs/journal/02-exploring-recent-events.md]: I'm leaning toward having a dedicated section/doc on the cold/warm/hot terminology.
I have a some draft content I'll put into another PR.

File: docs/journal/02-exploring-recent-events.md (hunk 6 above)
Comment (bennage) [inline docs/journal/02-exploring-recent-events.md]: Yes, we should mention this in the doc.

**Comment by MikeWasson** [comment]
@bennage  Updated based on feedback

**Pull request merged**